dataset/src/main/scala/frameless/TypedDatasetForwarded.scala

```diff
@@ -215,10 +215,25 @@ trait TypedDatasetForwarded[T] { self: TypedDataset[T] =>
     * Note that, this function is not a typical set union operation, in that it does not eliminate
     * duplicate items.  As such, it is analogous to `UNION ALL` in SQL.
     *
+    * Differs from `Dataset#union` by aligning fields if possible.
+    *
+    * Example:
+    * {{{
+    *   case class Foo(x: Int, y: Long)
+    *   case class Bar(y: Long, x: Int)
+    *   df1: TypedDataset[Foo] = ...
+    *   df2: TypedDataset[Bar] = ...
+    *
+    *   df1.union(df2): TypedDataset[Foo]
+    * }}}
+    *
     * apache/spark
     */
-  def union(other: TypedDataset[T]): TypedDataset[T] =
-    TypedDataset.create(dataset.union(other.dataset))
+  def union[T2](other: TypedDataset[T2]): TypedDataset[T] = {
```

Comment:
I don't like that we lose compile-time safety here, allowing to union with ```TypedDataset``` of different type. Is there any trick that could allow us to enforce it better in types? 

If not, I feel like it might be better to have two different functions and/or return something like ```Either```.

Reply:
Something like this:
``` scala
  def union[T2](other: TypedDataset[T2]): Option[TypedDataset[T]] = {
    val otherColumnTypes = other.dataset.dtypes.toMap

    val condition = dataset.dtypes.forall { case (key, value) =>
      otherColumnTypes.get(key).contains(value)
    }

    if(condition) {
      val columns = dataset.columns.map(new Column(_)).toSeq
      val alignedDataset = TypedDataset.createUnsafe[T](other.dataset.select(columns: _*))
      Some(TypedDataset.create(dataset.union(alignedDataset.dataset)))
    } else
      None
  }
```

Reply:
> I don't like that we lose compile-time safety here, allowing to union with TypedDataset of different type. Is there any trick that could allow us to enforce it better in types?

[maybe this](https://github.com/milessabin/shapeless/blob/8511884030ac061c6ab01940b188fcab53aff151/core/src/main/scala/shapeless/ops/hlists.scala#L1981-L2005)?

Reply:
@Avasil, what we do with SmartProject can also be relevant to this.

Reply:
Thanks for suggestions, I've never touched Shapeless before so I'm taking this opportunity to check it out. :)

Reply:
@OlivierBlanvillain @imarios 

Okay, I've came up with this:
```scala
def union[T2, R <: HList, R2 <: HList]
  (other: ReturnType[T2])
  (implicit
    g1: LabelledGeneric.Aux[T, R],
    g2: LabelledGeneric.Aux[T2, R2],
    A: Align[R, R2]
  ): ReturnType[T]
```
It works for ```case class``` but not when ```T``` is plain value like ```Int``` (because of ```LabelledGeneric```). I've asked in shapeless gitter channel if it is possible to have function like that accept both ```case classes``` and "normal" values and the recommendation was to create my own typeclass describing this operation.

Before I start doing it I'd like to ask what do you think about this approach (maybe you already have solution present in the codebase, so no need to reinvent wheel)

Reply:
@Avasil Looks good! Not sure if there is much do for `TypedDataset[Int]` given that the field name if not encoded in the type...

If you don't mind iterating just update your PR and we will comment on it. I don't think you are reinventing the wheel, but even if you did it's the best learning exercise :wink:

Reply:
@OlivierBlanvillain Sure, I'll try it, it might take me some time because my brain is still digesting Shapeless but it's super interesting and maybe I'll find a way to do it. :) 

Maybe it will be possible to get field name from ```StructField```, otherwise we could just settle for different method for non-case classes or something

Reply:
I was thinking that using the smart project type class we have in Frameless might be an easy way to code union. Say you want to union Datasets of type U and T. If you project U into T then you can easily union the two (as long as T has all the fields of U with the same names and types).

Maybe in the future we can have smarter union that will union even if the two dataset are not alignable. It will try to find a schema that combines both schema (like union the schemes first and then union the data). 

For now the solution using Smart a project should be good enough to give value.

Reply:
I don’t think that’s an issue. Just have an overloaded method that works on union same typed datasets.

Reply:
Well if that's ok and we don't care if names are a bit different for this case then it simplifies a lot.

The difference between ```SmartProject``` and ```Align``` is that ```SmartProject``` will compile union of:
```scala
final case class Foo[A, B](x: A, y: B)

final case class Bar[A, B, C](y: B, x: A, z: C)
```
and ```Align``` wil not. I was under assumption that it shouldn't compile but that might be wrong.

BTW Is it possible to make ```ImplicitNotFound``` annotation work for ```def```'s? Or do I need to do something similiar to ```SmartProject``` (case class + implicit def in companion object) if we go with ```Align```?

Reply:
Only works for traits and classes.